File: Python/min_cost_to_connect_all_points.py
import sys


def minCostConnectPoints_krushal(points: list) -> int:
    def find(index: int) -> int:
        if parent[index] != index:
            parent[index] = find(parent[index])
        return parent[index]

    def union(x: int, y: int) -> bool:
        root_x, root_y = find(x), find(y)
        if root_x == root_y:
            return False

        if rank[root_x] < rank[root_y]:
            root_x, root_y = root_y, root_x

        rank[root_x] += rank[root_y]
        parent[root_y] = root_x
        return True

    dist = lambda x, y: abs(points[x][0] - points[y][0]) + abs(points[x][1] - points[y][1])
    length = len(points)
    rank = [1] * length
    parent = list(range(length))
    edges = list()

    for i in range(length):
        for j in range(i + 1, length):
            edges.append((dist(i, j), i, j))

    edges.sort()

    ret, num = 0, 1
    for weight, x, y in edges:
        if union(x, y):
            ret += weight
            num += 1
            if num == length:
                break

    return ret


def minCostConnectPoints_greed(points: list) -> int:
    n, ans = len(points), 0
    seen = set()
    dis = [float('inf')] * n
    cur = 0
    
    for i in range(n-1):
        x, y = points[cur]
        seen.add(cur)
        for j, (nx, ny) in enumerate(points):
            if j in seen: continue
            dis[j] = min(dis[j], abs(nx-x)+abs(ny-y))
        s, cur = min((d, j) for j, d in enumerate(dis))
        dis[cur] = float('inf')
        ans += s
        
    return ans


if __name__ == "__main__":
    print(minCostConnectPoints_greed(
        [[0, 0], [2, 2], [3, 10], [5, 2], [7, 0]]))

File: core/tmp/Java/src/com/dbc/algorithm/UnionFindSets/MinCostToConnectAllPoints.java
package com.dbc.algorithm.UnionFindSets;

import java.util.*;

public class MinCostToConnectAllPoints {
    private int[] parent;
    private int[] rank;

    public int minCostConnectPoints_union_find(int[][] points) {
        int length = points.length;
        this.parent = new int[length];
        this.rank = new int[length];
        for (int i = 0; i < length; i++)
            parent[i] = i;

        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < length; i++) {
            for (int j = i + 1; j < length; j++) {
                int weight = Math.abs(points[i][0] - points[j][0]) + Math.abs(points[i][1] - points[j][1]);
                Edge edge = new Edge(weight, i, j);
                edges.add(edge);
            }
        }

        Collections.sort(edges, new Comparator<Edge>() {
            @Override
            public int compare(Edge o1, Edge o2) {
                return o1.weight - o2.weight;
            }
        });

        int result = 0, num = 1;
        for (Edge edge : edges) {
            if (union(edge.start, edge.end)) {
                result += edge.weight;
                num++;
                if (num == length)
                    break;
            }
        }

        return result;
    }

    public int find(int index) {
        if (index != this.parent[index]) {
            this.parent[index] = find(this.parent[index]);
        }
        return this.parent[index];
    }

    public boolean union(int index1, int index2) {
        int rootX = find(index1);
        int rootY = find(index2);
        if (rootX == rootY) {
            return false;
        }

        if (this.rank[rootX] < this.rank[rootY]) {
            int temp = rootX;
            rootX = rootY;
            rootY = temp;
        }
        this.rank[rootX] += this.rank[rootY];
        this.parent[rootY] = rootX;
        return true;
    }

    class Edge {
        private int weight;
        private int start;
        private int end;

        public Edge(int weight, int start, int end) {
            this.weight = weight;
            this.start = start;
            this.end = end;
        }
    }
}
